src/V3Split.cpp

```diff
@@ -794,21 +791,36 @@ class EmitSplitVisitor final : public VNVisitor {
 };
 
 class RemovePlaceholdersVisitor final : public VNVisitor {
-    std::unordered_set<AstNode*> m_removeSet;  // placeholders to be removed
-public:
-    explicit RemovePlaceholdersVisitor(AstNode* nodep) {
-        iterate(nodep);
-        for (AstNode* np : m_removeSet) {
-            np->unlinkFrBack();  // Without next
-            VL_DO_DANGLING(np->deleteTree(), np);
+    // MEMBERS
+    int m_emptyAlways{0};
+
+    // CONSTRUCTORS
+    RemovePlaceholdersVisitor() = default;
+    virtual ~RemovePlaceholdersVisitor() override = default;
+
+    // VISITORS
+    virtual void visit(AstSplitPlaceholder* nodep) override { pushDeletep(nodep->unlinkFrBack()); }
+    virtual void visit(AstNodeIf* nodep) override {
+        iterateChildren(nodep);
+        if (!nodep->ifsp() && !nodep->elsesp()) { pushDeletep(nodep->unlinkFrBack()); }
```

Comment:
This needs to check !isPureTreeRecurse(nodep->condp()) as the condition might have side effects.  However I don't think we want to do this here at all, I think it's better left in V3Const.

Reply:
Hmm, I am not sure if the pureness check is necessary.
This `RemovePlaceholdersVisitor` is used only against newly created always blocks.
https://github.com/verilator/verilator/blob/b9832171501761aad73fc84005af3743bd68deff/src/V3Split.cpp#L842-L850

If a node is not pure, the node is marked NotSplittable . So I think such nodes are not cloned and passed to `RemovePlaceholderVisitor` .
https://github.com/verilator/verilator/blob/b9832171501761aad73fc84005af3743bd68deff/src/V3Split.cpp#L424-L431

Reply:
Pureness check is added in de9a972663513eb05981b234dad9b13e235003f4